proj-android/PowerAuthLibrary/src/main/java/io/getlime/security/powerauth/networking/client/HttpRequestHelper.java

```diff
@@ -0,0 +1,363 @@
+            final PowerAuthAuthorizationHttpHeader header = helper.getAuthorizationHeader(available, requestData, requestMethod, endpoint.getAuthorizationUriId(), authentication);
+            if (header.getPowerAuthErrorCode() != PowerAuthErrorCodes.PA2Succeed) {
+                if (header.getPowerAuthErrorCode() == PowerAuthErrorCodes.PA2ErrorCodePendingProtocolUpgrade) {
+                    throw new PowerAuthErrorException(header.getPowerAuthErrorCode(), "Request is temporarily unavailable, due to pending protocol upgrade.");
+                }
+                throw new PowerAuthErrorException(header.getPowerAuthErrorCode());
+            }
+            // Keep authorization header
+            requestHeaders.put(header.getKey(), header.getValue());
+        }
+
+        // Finalize headers
+        requestHeaders.put("Content-Type", "application/json");
+        requestHeaders.put("Accept", "application/json");
+
+        // Return object with all information prepared for request processing.
+        return new RequestData(requestUrl, requestMethod, requestHeaders, requestData);
+    }
+
+
+    /**
+     * Build a response object from provided data. In case of error throws {@link FailedApiException},
+     * {@link ErrorResponseApiException} or {@link PowerAuthErrorException} exceptions.
+     *
+     * @param responseCode HTTP response status code
+     * @param responseData Response bytes
+     * @return object created from response bytes
+     * @throws Throwable if a deserialization, or decryption error occured.
+     */
+    @Nullable
+    public TResponse buildResponse(int responseCode, @Nullable byte[] responseData) throws Throwable {
+
+        if (responseCode / 100 != 2) {
+            // Non-200 response, throw an error
+            throw buildResponseException(responseCode, responseData, null, null);
+        }
+        try {
+            // 2xx response, try to build an object from response data
+            return buildResponseObject(responseCode, responseData);
+
+        } catch (FailedApiException | ErrorResponseApiException | PowerAuthErrorException e) {
+            // Known exceptions, just re-throw
+            throw e;
+        } catch (Throwable e) {
+            // In case of generic exception re-throw that
+            throw buildResponseException(responseCode, responseData, null, e);
+        }
+    }
+
+
+    /**
+     * Builds a response object from provided data. The function is used only for 2xx HTTP status codes,
+     * to process a success response object.
+     *
+     * @param responseCode HTTP response status code
+     * @param responseData Response bytes
+     * @return Response object, or null if response object is not specified.
+     * @throws Throwable if object cannot be constructed.
+     */
+    @Nullable
+    private TResponse buildResponseObject(int responseCode, @Nullable byte[] responseData) throws Throwable {
+
+        final byte[] objectData;
```

Comment:
Just a comment - declaring `byte[]` final will not make it immutable.

Reply:
The point for `final` here is to assign the value to the variable for only once. I do not modify byte arrays. Am I wrong? :)

Reply:
No, you are right :)